helm/revoke.go

```diff
@@ -0,0 +1,156 @@
+// See https://github.com/helm/helm/issues/4273
+func RevokeAccess(
+	kubectlOptions *kubectl.KubectlOptions,
+	tillerNamespace string,
+	rbacGroups []string,
+	rbacUsers []string,
+	serviceAccounts []string,
+) error {
+	logger := logging.GetProjectLogger()
+	logger.Infof(
+		"Revoking access Tiller server deployed in namespace %s to:",
+		tillerNamespace,
+	)
+	logEntities(rbacGroups, rbacUsers, serviceAccounts)
+
+	logger.Info("Checking if Tiller is deployed in the namespace.")
+	if err := validateTillerDeployed(kubectlOptions, tillerNamespace); err != nil {
+		logger.Errorf("Did not find a deployed Tiller instance in the namespace %s.", tillerNamespace)
+		return err
+	}
+	logger.Infof("Found a valid Tiller instance in the namespace %s.", tillerNamespace)
+
+	if len(rbacGroups) > 0 {
+		logger.Infof("Revoking access from RBAC groups")
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, convertToGroupInfos(rbacGroups)); err != nil {
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC groups")
+	} else {
+		logger.Infof("No RBAC groups to revoke - skipping")
+	}
+
+	if len(rbacUsers) > 0 {
+		logger.Infof("Revoking access from RBAC users")
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, convertToGroupInfos(rbacUsers)); err != nil {
+			return err
+		}
+		logger.Infof("Successfully revoked access from RBAC users")
+	} else {
+		logger.Infof("No RBAC users to revoke - skipping")
+	}
+
+	if len(serviceAccounts) > 0 {
+		logger.Infof("Revoking access from RBAC Service Accounts")
+		serviceAccountInfos, err := convertToServiceAccountInfos(serviceAccounts)
+		if err != nil {
+			return err
+		}
+		if err := revokeAccessFromRBACEntities(kubectlOptions, tillerNamespace, serviceAccountInfos); err != nil {
+			logger.Errorf("Error revoking access from Service Accounts: %s", err)
+			return err
+		}
+		logger.Infof("Successfully revoked access from Service Accounts")
+	} else {
+		logger.Infof("No Service Accounts to revoke - skipping")
+	}
+
+	return nil
+}
+
+// revokeAccessFromRBACEntities will revoke access to the deployed Tiller server to the provided RBAC groups.
+// The revocation process is as follows:
+// For each RBAC entity:
+// - Discover and remove the role and rolebinding
+// - Discover and remove the signed TLS keypair
+func revokeAccessFromRBACEntities(
+	kubectlOptions *kubectl.KubectlOptions,
+	tillerNamespace string,
+	rbacEntities []RBACEntity,
+) error {
+	logger := logging.GetProjectLogger()
+
+	numEntities := len(rbacEntities)
+	for idx, rbacEntity := range rbacEntities {
+		logger.Infof("Revoking access to entity %s (%d of %d entities)", rbacEntity, idx+1, numEntities)
+
+		logger.Infof("Deleting entity role")
+		if err := deleteEntityRoleAndBinding(kubectlOptions, tillerNamespace, rbacEntity); err != nil {
+			logger.Errorf("Unable to delete role for entity %s from namespace %s", rbacEntity.Subject().Name, tillerNamespace)
+			return err
```

Comment:
Might be better to collect all the errors, and then return a summary error so that all the groups in the list are attempted to be revoked before the one that fails.

Reply:
You can do similar things for the loop for other entities.

Reply:
For collecting the errors, do you propose to return `[]error` from `revokeAccessFromRBACEntities()` (in case more than one error was seen when trying to delete role and role binding) as well as `[]error` from `RevokeAccess()` (in case more than one error was seeing when trying to delete from user, group, and service account)? Or were you thinking just logging a warning and continuing on? I was modeling the current behavior from the grant function which returns error whenever it encounters one and doesn't collect many errors.

Reply:
Since a list of errors (`[]error`) is not actually an error in golang, you need to actually create a new type. See https://github.com/gruntwork-io/kubergrunt/blob/master/eks/errors.go#L47 for an example error type, and https://github.com/gruntwork-io/kubergrunt/blob/master/eks/instances.go#L51 for example usage.

> I was modeling the current behavior from the grant function which returns error whenever it encounters one and doesn't collect many errors.

Yup understood! This was an (unfortunately undocumented) improvement I've been thinking about. You know, the improved thought that comes to you when you see things a second time :) Anyway, feel free to just add a note with this thought so it isn't lost and leave it for a future improvement. Although would appreciate tackling it if you have time and since you're already in there!

Reply:
Thanks for the pointer. Nice to be able to model after an existing pattern so I don't deviate from the norm. I've updated this code.